### PR TITLE
Adaptive Tile Sampling

### PR DESCRIPTION
**Pull request by oktomus**
Introduce adaptive tile renderer.

### CODE REVIEW

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 1]
```diff
@@ -510,6 +515,7 @@ namespace
             m_image_plane_sampler_combo->setToolTip(m_params_metadata.get_path("sampling_mode.help"));
             m_image_plane_sampler_combo->addItem("Uniform", "uniform");
             m_image_plane_sampler_combo->addItem("Adaptive", "adaptive");
+            m_image_plane_sampler_combo->addItem("Adaptive Tile", "tile-adaptive");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Maybe `adaptive_tile` instead of `tile-adaptive`?

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp [hunk 2]
```diff
@@ -0,0 +1,521 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp]: Fix copyright notice. For files you created, there should be a single notice line with text

    Copyright (c) 2018 Kevin Masson, The appleseedhq Organization

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp (hunk 1 above)
Comment (oktomus) [reply src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Yes :)

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp (hunk 2 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp]: Ok :)

File: src/appleseed/foundation/image/filteredtile.cpp [hunk 3]
```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
...
+    FilteredTile*       main,
+    FilteredTile*       second)
+{
+    float error = 0.0;
+
+    assert(main->get_crop_window() == second->get_crop_window());
+    assert(main->get_crop_window().contains(bb.min));
+    assert(main->get_crop_window().contains(bb.max));
+    assert(bb.min.y >= 0 && bb.max.y <= main->get_height());
+    assert(bb.min.x >= 0 && bb.max.x <= main->get_width());
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.cpp]: Switch these two assertions (x first, y second). Are they actually necessary given the two assertions above?

File: src/appleseed/foundation/image/filteredtile.cpp [hunk 4]
```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
...
+        abs(main_color.r - second_color.r)
+        + abs(main_color.g - second_color.g)
+        + abs(main_color.b - second_color.b)
+       ) / fast_sqrt(main_color.r + main_color.g + main_color.b);
+}
+
+float FilteredTile::compute_tile_variance(
+    const AABB2u&       bb,
+    FilteredTile*       main,
+    FilteredTile*       second)
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.cpp]: The pointers must be const.

File: src/appleseed/foundation/image/filteredtile.h [hunk 5]
```diff
@@ -131,6 +137,20 @@ class FilteredTile
...
         const float*        values);
 
+    // Compute the variance of a weighted pixel given the value of the same pixel with 2 different samples count.
+    // The first given pixel `main` contains N samples.
+    // The second given pixel `second` contains N/2 samples (samples included in `second` are also in `main`).
+    static float compute_weighted_pixel_variance(
+        const float*        main,
+        const float*        second);
+
+    // Compute the variance of the tile `second` for pixels in the bouding box `bb`.
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.h]: > of the tile `second`
Are you sure it's not of the tile `main`?

Typo: bouding -> bounding

File: src/appleseed/foundation/image/filteredtile.cpp [hunk 6]
```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
     }
 }
 
+float FilteredTile::compute_weighted_pixel_variance(
+    const float*        main,
+    const float*        second)
+{
+    // Get weights.
+    float main_weight = *main++;
+    main_weight = main_weight == 0.0f ? 0.0f : 1.0f / main_weight;
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.cpp]: I suggest introducing a `rcp_main_weight` variable for clarity, and making all variables const.

File: src/appleseed/foundation/image/filteredtile.cpp [hunk 7]
```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
...
+    float second_weight = *second++;
+    second_weight = second_weight == 0.0f ? 0.0f : 1.0f / second_weight;
+
+    // Get colors and assign weights.
+    Color4f main_color(abs(main[0]), abs(main[1]), abs(main[2]), abs(main[3]));
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.cpp]: Make sure to include `<cmath>` otherwise you may have bad surprises with `std::abs()`!

File: src/appleseed/foundation/image/filteredtile.cpp [hunk 8]
```diff
@@ -163,4 +165,57 @@ void FilteredTile::atomic_add(
...
+
+    Color4f second_color(abs(second[0]), abs(second[1]), abs(second[2]), abs(second[3]));
+    second_color *= second_weight;
+
+    // Compute variance.
+    return (
+        abs(main_color.r - second_color.r)
+        + abs(main_color.g - second_color.g)
+        + abs(main_color.b - second_color.b)
+       ) / fast_sqrt(main_color.r + main_color.g + main_color.b);
```
Comment (dictoon) [inline src/appleseed/foundation/image/filteredtile.cpp]: At this point you could try multiplying by `fast_rcp_sqrt()`.

File: src/appleseed/foundation/image/filteredtile.cpp (hunk 7 above)
Comment (dictoon) [reply src/appleseed/foundation/image/filteredtile.cpp]: Actually not sure if computing absolute values here makes sense.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 9]
```diff
@@ -1786,6 +1786,90 @@ namespace
...
+      private:
+
+        // Merge frame::save_extra_aovs and PixelRenderer::enable_diagnostics.
+        void update_diagnostic_option()
+        {
+            for (each<ConfigurationContainer> i = m_project.configurations(); i; ++i)
+            {
+                Dictionary& root = i->get_parameters();
+
+                if (root.dictionaries().exist("uniform_pixel_renderer"))
```
[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Yes he has because of `invalid_samples` AOV

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 10]
```diff
@@ -1786,6 +1786,90 @@ namespace
...
+        void update_pixel_renderer_option()
+        {
+            for (each<ConfigurationContainer> i = m_project.configurations(); i; ++i)
+            {
+                Dictionary& root = i->get_parameters();
+
+                if (!root.dictionaries().exist("final"))
+                    continue;
+
+                Dictionary& gtr = root.dictionary("final");
```
[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: It works I tested but you're right, it makes no sens to do it that way.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 11]
```diff
@@ -1786,6 +1786,90 @@ namespace
...
+                // The frame does not reference any camera: use the first camera.
+                frame_params.insert(
+                    "enable_diagnostic_aovs",
+                    frame_params.get<bool>("save_extra_aovs"));
+
+                frame_params.strings().remove("save_extra_aovs");
+            }
+        }
+
+        // Rename pixel_renderer to sampling_method.
```
[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Because if we do `adaptive_tile` we don't use any `PixelRenderer`.

File: src/appleseed/foundation/image/filteredtile.cpp (hunk 7 above)
Comment (oktomus) [reply src/appleseed/foundation/image/filteredtile.cpp]: It happened that some shading values are negative and makes the metric completly wrong

Comment (oktomus) [reply src/appleseed/foundation/image/filteredtile.cpp]: Most of the times it was because of the filter that has negative weights

File: src/appleseed/foundation/image/filteredtile.cpp (hunk 3 above)
Comment (oktomus) [reply src/appleseed/foundation/image/filteredtile.cpp]: No :grimacing:

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: I don't think it can work since `root.dictionaries().exist("final")` will always be false and thus the body loop will never be executed...

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: > Because if we do adaptive_tile we don't use any PixelRenderer.

I don't understand how that's an argument to rename `pixel_renderer` to `sampling_method`. In all other cases we do have pixel renderers.

Let's revert this.

[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: I find this weird to name it pixel_renderer because if `pixel_renderer = adaptive_tile` we don't use any pixel renderer..

[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: I don't follow. AdaptiveTileRenderer is a tile renderer so it's the `tile_renderer` parameter that must be set to `adaptive_tile_renderer` and `pixel_renderer` must not be set at all.

[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: The hierarchy of renderer components (frame/tile/pixel/sample) described in https://github.com/appleseedhq/appleseed/wiki/Renderer-Components is fixed. You simply plug yourself at the tile level instead of at the pixel level.

[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: If you imagine the user behavior maybe you will see my point of view. When you are in the ui, chossing for the renderer, you have only one dropdown that switches between `uniform`, `adaptive` and `adaptive_tile`. For me this is the method and depending on the method, we choose correct pixel and tile renderers

[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: From a user's point of view, the hierarchy of frame/tile/pixel/sample renderers is clearly irrelevant. What we do under the hood is separate from what how we present things to the user.

[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: So the problem is in the UI. I should not link the dropdown to `pixel_renderer` but instead do a regular dropdown and edit `pixel_renderer` or `tile_renderer` accordingly.

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp [hunk 12]
```diff
@@ -0,0 +1,897 @@
...
+                m_samples_aov_index = frame.create_extra_aov_image("samples");
+
+                if ((thread_index == 0) && m_samples_aov_index == ~size_t(0))
+                {
+                    RENDERER_LOG_WARNING(
+                        "could not create some of the diagnostic aovs, maximum number of aovs (" FMT_SIZE_T ") reached.",
+                        MaxAOVCount);
+                }
+            }
+
```
Comment (est77) [inline src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp]: Remove extra empty line

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp [hunk 13]
```diff
@@ -0,0 +1,897 @@
...
+                    child_sampling_context,
+                    pixel_context,
+                    sample_position,
+                    m_aov_accumulators,
+                    shading_result);
+
+                // Ignore invalid samples.
+                if (!shading_result.is_valid())
+                {
+                    //signal_invalid_sample();
```
Comment (est77) [inline src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp]: Remove commented code

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Exactly.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 14]
```diff
@@ -1786,6 +1786,60 @@ namespace
...
+            for (each<ConfigurationContainer> i = m_project.configurations(); i; ++i)
+            {
+                Dictionary& root = i->get_parameters();
+
+                if (root.dictionaries().exist("uniform_pixel_renderer"))
+                {
+                    Dictionary& upr = root.dictionary("uniform_pixel_renderer");
+
+                    if (upr.strings().exist("enable_diagnostics"))
+                        upr.strings().remove("enable_diagnostics");
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: You can actually call `remove()` without first checking if the key exists; it won't do anything if the key doesn't exist.

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 15]
```diff
@@ -1006,7 +1011,7 @@ void Frame::extract_parameters()
     }
 
     // Retrieve save extra AOVs parameter
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Update comment (and add a period at the end).

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 16]
```diff
@@ -364,6 +365,10 @@ bool RendererComponents::create_pixel_renderer_factory()
 
         return true;
     }
+    else if (name == "adaptive_tile")
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: That doesn't make sense: `pixel_renderer` will simply not be set when using the adaptive tile renderer. You must remove this branch.

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 17]
```diff
@@ -443,6 +467,10 @@ bool RendererComponents::create_tile_renderer_factory()
         m_tile_renderer_factory.reset(new DebugTileRendererFactory());
         return true;
     }
+    else if (name.empty())
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: Keep the empty case first like it was done before, and like it's done in other similar functions.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp [hunk 18]
```diff
@@ -115,6 +109,7 @@ void PixelRendererBase::on_tile_end(
 
                 switch (sample_state[0])
                 {
+                  default:
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Why this change? What other sample states can there be?

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h [hunk 19]
```diff
@@ -53,7 +53,7 @@ namespace renderer
 //
 
 class UniformPixelRendererFactory
-  : public PixelRendererBaseFactory
+  : public IPixelRendererFactory
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h]: Nobody is inheriting from `PixelRendererBase` anymore? That's unexpected. The point of `PixelRendererBase` was to provide default (empty) implementations for a number of methods. However if truly no pixel renderer needs it then it needs to be removed.

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.h [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_RENDERER_KERNEL_RENDERING_FINALE_ADAPTIVETILERENDERER_H
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.h]: Typo: `FINALE` -> `FINAL`. Don't forget to fix it at the end of the file too.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 21]
```diff
@@ -485,13 +487,45 @@ namespace
...
+
+            // Set the pixel and tile renderer.
+            const QString sampler =
+                m_image_plane_sampler_combo->itemData(
+                    m_image_plane_sampler_combo->currentIndex()
+                ).value<QString>();
+
+            if (sampler == "adaptive_tile")
+            {
+                set_config(
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Make those statements fit on a single line? :)

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 22]
```diff
@@ -586,10 +618,65 @@ namespace
...
+            adaptiveness->setToolTip(m_params_metadata.get_path("adaptive_tile_renderer.adaptiveness.help"));
+            sublayout->addRow("Adaptiveness", adaptiveness);
+        }
+
+        void load_general_sampler(const Configuration& config)
+        {
+            const string default_tr_path = "tile_renderer.default";
+            const string default_tr_value =
+                m_params_metadata.get_path_optional<string>(
+                        default_tr_path.c_str(), "");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Fix indentation (or move to the previous line).

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 23]
```diff
@@ -586,10 +618,65 @@ namespace
...
+            if (tr_value == "adaptive")
+            {
+                m_image_plane_sampler_combo->setCurrentIndex(2);
+                return;
+            }
+
+            const string default_pr_path = "pixel_renderer.default";
+            const string default_pr_value =
+                m_params_metadata.get_path_optional<string>(
+                        default_pr_path.c_str(), "");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Fix indentation (or move to the previous line).

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp (hunk 16 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: I forgot sry.

File: src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp (hunk 18 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp]: Because I forgot to clear the tile before all. It will be removed when the AOV branch will be merged.

File: src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h (hunk 19 above)
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h]: It's just the factory. It was like this before I implemented the save_diagnostic PR months ago.

Comment (dictoon) [reply src/appleseed/renderer/kernel/rendering/final/uniformpixelrenderer.h]: Oh right, thanks.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 24]
```diff
@@ -483,13 +487,28 @@ namespace
...
                 get_widget<size_t>("general.passes") > 1 ? "permanent" : "ephemeral");
+
+            // Set the pixel and tile renderer.
+            const QString sampler = m_image_plane_sampler_combo->itemData(
+                m_image_plane_sampler_combo->currentIndex()).value<QString>();
+
+            set_config(
+                config,
+                "pixel_renderer",
+                sampler == "adaptive_tile" ? "" : sampler.toLatin1().data());
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: `toAscii()` rather.

File: src/appleseed/renderer/modeling/aov/diagnosticaov.cpp [hunk 25]
```diff
@@ -165,15 +164,15 @@ namespace
 
         void on_pixel_end(const Vector2i& pi) override
         {
-            // Store a hint corresping to the sample state in the tile.
+            // Store a hint correspoding to the sample state in the tile.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/diagnosticaov.cpp]: Still a typo there :)

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 26]
```diff
@@ -75,6 +76,9 @@ namespace
         const ParamArray&   source,
         const char*         param_name)
     {
+        std::cout << "Copygin param " << param_name << "\n";
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: Remove debug code.

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 27]
```diff
@@ -340,7 +345,7 @@ bool RendererComponents::create_pixel_renderer_factory()
         }
 
         ParamArray params = get_child_and_inherit_globals(m_params, "uniform_pixel_renderer");
-        copy_param(params, get_child_and_inherit_globals(m_params, "generic_frame_renderer"), "passes");
+        copy_param(params, m_params, "passes");
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: Redundant with what `get_child_and_inherit_globals()` do.

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 28]
```diff
@@ -426,8 +431,7 @@ bool RendererComponents::create_tile_renderer_factory()
         }
 
         ParamArray params = get_child_and_inherit_globals(m_params, "adaptive_tile_renderer");
-        copy_param(params, get_child_and_inherit_globals(m_params, "generic_frame_renderer"), "passes");
-
+        copy_param(params, m_params, "passes");
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: Same problem.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 29]
```diff
@@ -1834,6 +1835,29 @@ namespace
             ParamArray& frame_params = frame->get_parameters();
             frame_params.strings().remove("save_extra_aovs");
         }
+
+        // Move generic_frame_renderer::passes to the root configuration.
+        void update_passes_path()
+        {
+            for (each<ConfigurationContainer> i = m_project.configurations(); i; ++i)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: This is perfectly fine and you can leave it like that. Note however that in new code we preer to use C++11's range-based `for` loops:

    for (Configuration& config :  m_project.configurations())

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 30]
```diff
@@ -1834,6 +1835,29 @@ namespace
...
+                {
+                    Dictionary& gfr = root.dictionary("generic_frame_renderer");
+
+                    if (gfr.strings().exist("passes"))
+                        root.strings().insert("passes", gfr.get("passes"));
+
+                    gfr.strings().remove("passes");
+
+                    // Remove the dictionnary from the root if it's empty
+                    if (gfr.size() == 0)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Prefer `empty()` to `size() == 0`.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 31]
```diff
@@ -1834,6 +1835,29 @@ namespace
...
+        {
+            for (each<ConfigurationContainer> i = m_project.configurations(); i; ++i)
+            {
+                Dictionary& root = i->get_parameters();
+
+                if (root.dictionaries().exist("generic_frame_renderer"))
+                {
+                    Dictionary& gfr = root.dictionary("generic_frame_renderer");
+
+                    if (gfr.strings().exist("passes"))
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Use `move_if_exist()` that already does all this.

File: src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp [hunk 32]
```diff
@@ -0,0 +1,928 @@
...
+
+    //
+    // Adaptive tile renderer.
+    //
+
+#ifndef NDEBUG
+
+    // Define this symbol to break execution into the debugger
+    // when a specific pixel is about to be rendered.
+    // #define DEBUG_BREAK_AT_PIXEL Vector2i(0, 0)
```
[the remark this answers is not in the document]
Comment (oktomus) [reply src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp]: yes!

File: src/appleseed/renderer/modeling/aov/diagnosticaov.cpp (hunk 25 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/aov/diagnosticaov.cpp]: huh, sorry :(

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp (hunk 29 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: I prefer too.

File: src/appleseed/renderer/modeling/input/texturesource.cpp [hunk 33]
```diff
@@ -36,6 +36,7 @@
 #include "renderer/modeling/texture/texture.h"
 
 // appleseed.foundation headers.
+#include "foundation/image/color.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/texturesource.cpp]: In a future PR: you can remove this as it's already included by `texturesource.h`.

File: src/appleseed/renderer/modeling/project/configuration.cpp [hunk 34]
```diff
@@ -274,6 +281,8 @@ auto_release_ptr<Configuration> BaseConfigurationFactory::create_base_interactiv
     parameters.insert("spectrum_mode", "rgb");
     parameters.insert("sampling_mode", "qmc");
 
+    parameters.insert("passes", 1);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/configuration.cpp]: In a future PR: concept of passes doesn't make sense for interactive rendering; remove this.

Comment (oktomus) [reply src/appleseed/renderer/modeling/project/configuration.cpp]: ok!

File: src/appleseed/renderer/modeling/input/texturesource.cpp (hunk 33 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/input/texturesource.cpp]: done

File: src/appleseed/renderer/modeling/project/configuration.cpp (hunk 34 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/project/configuration.cpp]: done

Comment (dictoon) [reply src/appleseed/renderer/modeling/project/configuration.cpp]: Done where?

Comment (oktomus) [reply src/appleseed/renderer/modeling/project/configuration.cpp]: in the resumable render PR
